### PR TITLE
Bug : Fix date indexing

### LINKED ISSUE #433: Rethink indexing date time with precision
**Describe the issue to be researched**
Currently the in date index table we index the start and end datetime values as well as the precision.

**Describe the goal of the research**
Decide if the precision column is actually useful. I have the feeling that we probably should just index different start/end values depending on the precision at the point of indexing resources. Rather than keeping the precision column itself.

**Describe the methodology**
Understand the FHIR search standard esp. the date part: https://www.hl7.org/fhir/search.html#date
Research the internal representation of the HAPI FHIR lib we use

**Would you like to work on the issue?**
Happy to collab.

### PR DESCRIPTION
**Pull request by epicadk**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #433 

**Description**
Clear and concise code change description.  
Added precision to ts_high in the date index to comply with the standard. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR? 

We could index precision however I don't really see a usecase. 

**Type**
Choose one: Bug fix 

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [ ] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: core/src/main/java/com/google/android/fhir/index/entities/DateIndex.kt [hunk 1]
```diff
@@ -31,10 +29,5 @@ internal data class DateIndex(
   /** The high timestamp. */
   val tsHigh: Long,
   /** The low timestamp. */
-  val tsLow: Long,
-  /**
-   * The smallest value we can unambiguously resolve the date to. This is an indication to clients
-   * that any part of the timestamp smaller than the [temporalPrecision] should be ignored.
-   */
-  val temporalPrecision: TemporalPrecisionEnum
+  val tsLow: Long
```
Comment (jingtang10) [inline core/src/main/java/com/google/android/fhir/index/entities/DateIndex.kt:32]: while we're at it, can we change the order of these two fields and rename them `from` and `to`?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by jingtang10** [comment]
thanks for the draft! @epicadk

File: core/src/main/java/com/google/android/fhir/db/impl/entities/DateIndexEntity.kt [hunk 2]
```diff
@@ -27,7 +27,7 @@ import org.hl7.fhir.r4.model.ResourceType
 @Entity(
   indices =
     [
-      Index(value = ["resourceType", "index_name", "index_tsHigh", "index_tsLow"]),
+      Index(value = ["resourceType", "index_name", "index_to", "index_from"]),
```
Comment (jingtang10) [inline core/src/main/java/com/google/android/fhir/db/impl/entities/DateIndexEntity.kt:30]: can you change the order here too?

File: core/src/main/java/com/google/android/fhir/index/entities/DateIndex.kt [hunk 3]
```diff
@@ -29,12 +27,7 @@ internal data class DateIndex(
...
-  val tsHigh: Long,
+  val to: Long,
   /** The low timestamp. */
-  val tsLow: Long,
-  /**
-   * The smallest value we can unambiguously resolve the date to. This is an indication to clients
-   * that any part of the timestamp smaller than the [temporalPrecision] should be ignored.
-   */
-  val temporalPrecision: TemporalPrecisionEnum
+  val from: Long
```
Comment (jingtang10) [inline core/src/main/java/com/google/android/fhir/index/entities/DateIndex.kt:32]: change the order of these two fields here and everywhere in other files

add descriptive documentation and espeically describe whether these values are exclusive or inclusive.

File: core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt [hunk 4]
```diff
@@ -98,9 +98,8 @@ internal object ResourceIndexer {
         DateIndex(
           name = "_lastUpdated",
           path = arrayOf(resource.fhirType(), "meta", "lastUpdated").joinToString(separator = "."),
-          tsHigh = lastUpdatedElement.value.time,
-          tsLow = lastUpdatedElement.value.time,
-          temporalPrecision = lastUpdatedElement.precision
+          to = lastUpdatedElement.value.time,
+          from = lastUpdatedElement.precision.add(lastUpdatedElement.value, 1).time - 1
```
Comment (jingtang10) [inline core/src/main/java/com/google/android/fhir/index/ResourceIndexer.kt:102]: change order

**Review by jingtang10 — CHANGES_REQUESTED** [review]
Thanks!

**Comment by epicadk** [comment]
Done @jingtang10

**Review by jingtang10 — APPROVED** [review]

**Comment by jingtang10** [comment]
Thanks @epicadk for the contribution! 🎉

**Pull request merged**